include/oneapi/dpl/internal/dynamic_selection_traits.h

```diff
@@ -0,0 +1,220 @@
+  auto has_submit_and_wait_impl(int) -> decltype(std::declval<DSPolicy>().submit_and_wait(std::declval<Function>(), std::declval<Args>()...), std::true_type{});
+
+  template<typename DSPolicy, typename Function, typename... Args>
+  struct has_submit_and_wait : decltype(has_submit_and_wait_impl<DSPolicy, Function, Args...>(0)) {};
+
+  template<typename DSPolicy, typename SelectionHandle,  typename Function, typename... Args>
+  auto has_submit_and_wait_handle_impl(...) -> std::false_type;
+
+  template<typename DSPolicy, typename SelectionHandle,  typename Function, typename... Args>
+  auto has_submit_and_wait_handle_impl(int) -> decltype(std::declval<DSPolicy>().submit_and_wait(std::declval<SelectionHandle>(),
+                                                                                                 std::declval<Function>(),
+                                                                                                 std::declval<Args>()...), std::true_type{});
+
+  template<typename DSPolicy, typename SelectionHandle, typename Function, typename... Args>
+  struct has_submit_and_wait_handle : decltype(has_submit_and_wait_handle_impl<DSPolicy, SelectionHandle , Function, Args...>(0)) {};
+
+  template<typename S, typename Info>
+  auto has_report_impl(...) -> std::false_type;
+
+  template<typename S, typename Info>
+  auto has_report_impl(int) -> decltype(std::declval<S>().report(std::declval<Info>()), std::true_type{});
+
+  template<typename S, typename Info>
+  struct has_report : decltype(has_report_impl<S,Info>(0)) {};
+
+  template<typename S, typename Info>
+  auto has_report_value_impl(...) -> std::false_type;
+
+  template<typename S, typename Info>
+  auto has_report_value_impl(int) -> decltype(std::declval<S>().report(std::declval<Info>(), 0), std::true_type{});
+
+  template<typename S, typename Info>
+  struct has_report_value : decltype(has_report_value_impl<S,Info>(0)) {};
+
+ } //namespace internal
+
+  constexpr int deferred_initialization = -1;
+
+  // required interfaces
+  template<typename DSPolicy>
+  auto get_resources(DSPolicy&& dp) {
+    return std::forward<DSPolicy>(dp).get_resources();
+  }
+
+  template<typename WaitObject>
+  auto wait(WaitObject&& w) {
+    return std::forward<WaitObject>(w).wait();
+  }
+
+  template<typename DSPolicy>
+  auto  get_submission_group(DSPolicy&& dp){
+    return std::forward<DSPolicy>(dp).get_submission_group();
+  }
+
+  template<typename DSPolicy, typename... Args>
+  typename policy_traits<DSPolicy>::selection_type select(DSPolicy&& dp, Args&&... args) {
+    return std::forward<DSPolicy>(dp).select(std::forward<Args>(args)...);
+  }
+
+  // optional interfaces
+
+  template<typename T>
+  auto unwrap(T&& v) {
+    if constexpr(internal::has_unwrap<T>::value) {
+        return std::forward<T>(v).unwrap();
+    } else {
+        return v;
+    }
+  }
+
+  template<typename T, typename Function, typename... Args>
+  auto submit(T&& t, Function&&f, Args&&... args) {
```

Comment:
We agreed on not doing stylish fixes but please do this. Otherwise, readability suffers

```suggestion
  auto submit(T&& t, Function&& f, Args&&... args) {
```

Reply:
Not yet done.